ignite/metrics/maximum_mean_discrepancy.py

```diff
@@ -0,0 +1,131 @@
+from typing import Callable, Sequence
+
+import torch
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric, reinit__is_reduced, sync_all_reduce
+
+__all__ = ["MaximumMeanDiscrepancy"]
+
+
+class MaximumMeanDiscrepancy(Metric):
+    r"""Calculates the mean of `maximum mean discrepancy (MMD)
+    <https://www.onurtunali.com/ml/2019/03/08/maximum-mean-discrepancy-in-machine-learning.html>`_.
+
+    .. math::
+       \begin{align*}
+           \text{MMD}^2 (P,Q) &= \underset{\| f \| \leq 1}{\text{sup}} | \mathbb{E}_{X\sim P}[f(X)]
+           - \mathbb{E}_{Y\sim Q}[f(Y)] |^2 \\
+           &\approx \frac{1}{B(B-1)} \sum_{i=1}^B \sum_{\substack{j=1 \\ j\neq i}}^B k(\mathbf{x}_i,\mathbf{x}_j)
+           -\frac{2}{B^2}\sum_{i=1}^B \sum_{j=1}^B k(\mathbf{x}_i,\mathbf{y}_j)
+           + \frac{1}{B(B-1)} \sum_{i=1}^B \sum_{\substack{j=1 \\ j\neq i}}^B k(\mathbf{y}_i,\mathbf{y}_j)
+       \end{align*}
+
+    where :math:`B` is the batch size, and :math:`\mathbf{x}_i` and :math:`\mathbf{y}_j` are
```

Comment:
Sorry to coming back to this question, actually, I'm not quite sure about that B is the batch size.
According to https://www.onurtunali.com/ml/2019/03/08/maximum-mean-discrepancy-in-machine-learning.html, their `m` is the num of samples.
Also, https://jmlr.org/papers/volume13/gretton12a/gretton12a.pdf sets the problem as 
> Let x and y be random variables defined on a topological space X, with respective
Borel probability measures p and q . Given observations X := {x1, . . . , xm} and Y := {y1, . . . , yn},
independently and identically distributed (i.i.d.) from p and q, respectively, can we decide whether
p != q?

where m and n are num of samples of the observations.

Should not we use the same idea and replace B per N = num of samples ?

Reply:
Thank you for the suggestion.

Although it is ideal to compute MMD using all $N$ samples, doing so requires to compute $N^2$ combinations of k(x, y), where we have to pool all inputs.
So, this metric computes MMD for each batch and takes the average.

I think using $B$ would be better so that users can be aware of this point.